examples/dnn_yolo_train_ex.cpp

```diff
@@ -0,0 +1,526 @@
+        cout << " (" << (100.0*label.second)/num_objects << "%)\n";
+        options.labels.push_back(label.first);
+        string_to_color(label.first);
+    }
+
+    // If the default anchor boxes don't fit your data well you should recompute them.
+    // Here's an simple way to do it using K-Means clustering.  Note that the approach
+    // shown below is suboptimal, since it doesn't group the bounding boxes by size.
+    // Grouping the bounding boxes by size and computing the K-Means on each group
+    // would make more sense, since each stride of the network is meant to output
+    // boxes at a particular size, but that is very specific to the network architecture
+    // and the dataset itself.
+    if (parser.option("anchors"))
+    {
+        const auto num_clusers = std::stoul(parser.option("anchors").argument());
+        std::vector<dpoint> samples;
+        // First we need to rescale the bounding boxes to match the image size at training time.
+        for (const auto& image_info : dataset.images)
+        {
+            const auto scale = image_size / std::max<double>(image_info.width, image_info.height);
+            for (const auto& box : image_info.boxes)
+            {
+                dpoint sample(box.rect.width(), box.rect.height());
+                samples.push_back(sample*scale);
+            }
+        }
+        // Now we can compute K-Means clustering
+        randomize_samples(samples);
+        cout << "Computing anchors for " << samples.size() << " samples" << endl;
+        std::vector<dpoint> anchors;
+        pick_initial_centers(num_clusers, anchors, samples);
+        find_clusters_using_kmeans(samples, anchors);
+        std::sort(anchors.begin(), anchors.end(), [](const dpoint& a, const dpoint& b){ return prod(a) < prod(b); });
+        for (const dpoint& c : anchors)
+            cout << round(c(0)) << 'x' << round(c(1)) << endl;
+        // And check the average IoU of the newly computed anchor boxes and the training samples.
+        double average_iou = 0;
+        for (const dpoint& s : samples)
+        {
+            drectangle sample = centered_drect(dpoint(0, 0), s.x(), s.y());
+            double best_iou = 0;
+            for (const dpoint& a : anchors)
+            {
+                drectangle anchor = centered_drect(dpoint(0, 0), a.x(), a.y());
+                best_iou = std::max(best_iou, box_intersection_over_union(sample, anchor));
+            }
+            average_iou += best_iou;
+        }
+        cout << "Average IoU: " << average_iou / samples.size() << endl;
+        return EXIT_SUCCESS;
+    }
+
+    // When computing the objectness loss in YOLO, predictions that do not have an IoU
+    // with any ground truth box of at least options.iou_ignore_threshold, will be
+    // treated as not capable of detecting an object, an therefore incur loss.
+    // Predictions above this threshold will be ignored, i.e. will not contribute to the
+    // loss. Good values are 0.7 or 0.5.
```

Comment:
```suggestion
    // Similarly, predictions above this threshold are considered correct predictions by the loss. Typical settings for this threshold are in the range 0.5 to 0.7.
```
Since they aren't ignored.  The best matching ones still contribute to the loss.

Reply:
Yes, they are just ignored in this part of the loss (when computing the ones that should not detect an object).

But you are right, some of them (only the one that whose center falls into the same grid cell as the truth box) will still contribute to the loss in the second part (when we compute the predictions that should detect an object, location, and class).

The wording comes from the original YOLO implementation, so people familiar with the original code, can feel right at home when reading this.